keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -233,7 +233,589 @@ yet but we think may be interesting to consider in the future.
+- A number of queues; and
+- A queue length limit.
+
+Each non-exempt RequestPriority with more than one queue also has:
+- A hand size (a small positive number).
+
+The best matching FlowSchema for a given request is one of those whose
+boolean test accepts the request.  It is a configuration error if
+there is no FlowSchema that matches every request.  In case multiple
+schemas accept the request, the best is one of those with the
+logically highest matching priority.  In case there are multiple of
+those the implementation is free to pick any of those as best.  A
+matching priority is an integer, and a numerically lower number
+indicates a logically higher priority.
+
+A FlowSchema’s boolean test is constructed from atomic tests.  Each
+atomic test compares an authenticated request attribute --- selected
+from _either_ the client identity attributes or those that
+characterize the work being requested --- with a literal value
+(scalar, pattern, or set).  For every available atomic test, its
+inverse is also available.  Atomic tests can be ANDed together.  Those
+conjunctions can then be ORed together.  The predicate of a FlowSchema
+is such a disjunction.
+
+A FlowSchema’s rule for computing the request’s flow distinguisher
+identifies a string attribute of the authenticated request and
+optionally includes a transformation.  The available string attributes
+are (1) namespace of a resource-oriented request (available only if
+the predicate accepts only resource-oriented requests) and (2)
+username.  If no transformation is indicated then the flow
+distinguisher is simply the selected request attribute.  There is only
+one transformation available, and it is based on a regex that is
+configured in the flow schema and contains a capturing group.  The
+transformation consists of doing a complete match against the regex
+and extracting submatch number 1; if the selected string does not
+match the regex then the transformation yields the empty string.
+
+### Assignment to a Queue
+
+A non-exempt RequestPriority object also has a number of queues (we
+are talking about a number here, not the actual set of queues; the
+queues exist independently at each apiserver).  If the
+RequestPriority’s number of queues is more than one then the following
+logic is used to assign a request to a queue.
+
+For a given priority at a given apiserver, each queue is identified by
+a numeric index (starting at zero).  A RequestPriority has a hand size
+H (so called because the technique here is an application of shuffle
+sharding), a small positive number.  When a request arrives at an
+apiserver the request flow identifier’s string pair is hashed and the
+hash value is used to shuffle the queue indices and deal a hand of
+size H, as follows.  We use a hash function that produces at least 64
+bits, and 64 of those bits are taken as an unsigned integer we will
+call V.  The next step is finding the unique set of integers A[0] in
```

Comment:
For the shuffle sharding implementation, the method Daniel referenced in his [BucketClasses document](https://docs.google.com/a/google.com/document/d/1BtwFyB6G3JgYOaTxPjQD-tKHXaIYTByKlwlqLE0RUrk/edit?disco=AAAAC5efvXU) seems a bit easier to understand and possibly implement in my opinion.  Does that method seem equally sufficient and if so would it make sense to reference here?

Reply:
That technique is a bit simpler and suffers a consequence: there can be duplicate values in a hand.  Continuing the example from that doc, where a hand consists of eight 6-bit indices: if the hash value is 0x041041041041 then the hand contains eight copies of index 1.

Reply:
The technique outlined here has a simple recursive implementation.  The shuffle sharding can be accomplished by the function call `queueIndex := shuffleDealAndPick(hashValue, numQueues, func(a int) int {return a}, handSize, maxInt, -1)` where
```
type picker func(int) int

func shuffleDealAndPick(v, nq int64, p picker, nRem, minLen, bestIdx int) int {
    if nRem < 1 {
        return bestIdx
    }
    vNext := v/nq
    ai := int(v - nq*vNext)
    ii := p(ai)
    pNext := func(a int) int {
        if a < ai {
            return p(a)
        }
        return p(a+1)
    }
    lenI := lengthOfQueue(ii)
    if lenI < minLen {
        minLen = lenI
        bestIdx = ii
    }
    return shuffleDealAndPick(vNext, nq-1, pnext, nRem-1, minLen, bestIdx)
}
```

Reply:
https://play.golang.org/p/Ejb56vVnV_W

Clever :) 64 bits is probably not quite enough, though.

Reply:
It's easy to get more bits, we can just XOR the next N bits with the top N bits of `v`, where N is something like log2(nq). That's easier than doing multi-word division :)

Reply:
Yes, we could pull in more bits of v.  But I doubt we actually will need a large hand size with a large number of queues (i.e., large falling factorial).

BTW, `rand.Uint64()` is a thing.

Reply:
For a slightly more elaborate version of that recursion, with the domain and range of the mapping function commented and with debug printing, see https://play.golang.org/p/61J3CaTARGI

Reply:
I think it would be good in a followup to capture the code samples as e.g. footnotes or appendices. I think they will be easier for many to read than the textual description.